File: AppExamples/Flutter/lib/network/Repository.dart
import 'package:http/http.dart' as http;
import 'package:device_info_plus/device_info_plus.dart';

const BASE_URL = "http://192.168.1.68:3000/";
// change the ip above to the ebserver machine ip
const START_ENDPOINT = "what_now/";
const LOGDATA_ENDPOINT = "logdata/";
const FINISH_ENDPOINT = "done/";

const APP_ID = "com.example.starter";
const FRAMEWORK = "FLUTTER";
const TEST_TYPE = "benchmarkGame";

class ProgramExecution {
  late String program;
  late String parameter;

  ProgramExecution (this.program, this.parameter);
}

Future<ProgramExecution> start() async {
  AndroidDeviceInfo androidInfo = await DeviceInfoPlugin().androidInfo;
  var response = await http.get(
      Uri.parse(BASE_URL + START_ENDPOINT),
      headers: {
        'device': androidInfo.model!,
        'test_type': TEST_TYPE
      }
  );
  if (response.statusCode == 200){
    final splited = response.body.toString().split("-");
    return ProgramExecution(splited[0], splited[1]);
  }
  throw Exception("Error on starting code");
}

Future<void> logdata() async {
  AndroidDeviceInfo androidInfo = await DeviceInfoPlugin().androidInfo;
  var response = await http.get(
      Uri.parse(BASE_URL + LOGDATA_ENDPOINT),
      headers: {
        'device': androidInfo.model!,
        'test_type': TEST_TYPE,
        'application_id': APP_ID,
        'framework': FRAMEWORK
      }
  );
  if (response.statusCode != 200){
    throw Exception("Error on logging battery status code");
  }
}

void done() async {
  AndroidDeviceInfo androidInfo = await DeviceInfoPlugin().androidInfo;
  http.get(
      Uri.parse(BASE_URL + FINISH_ENDPOINT),
      headers: {
        'device': androidInfo.model!,
        'test_type': TEST_TYPE,
        'application_id': APP_ID,
        'activity': '.MainActivity'
      }
  );
}

File: Repositories/ReactNative/Repository.js
import type { Node } from 'react';
import DeviceInfo from 'react-native-device-info';

const APP_ID = "com.react.example.starter";
const FRAMEWORK = "REACT";
const TEST_TYPE = "example";
const ACTIVITY = '.MainActivity'
let hasCalled = false;

const localhostip = '192.168.1.68';

const App: () => Node = () => {
  useKeepAwake();
  let benchmark;
  const [isLoading, setLoading] = useState(true);

  const whatNow = async () => {
    if (hasCalled) { return; }
    console.log('http://' + localhostip + ':3000/' + 'what_now');
    try {
      let deviceModel = await DeviceInfo.getModel()
      const response = await fetch(
        'http://' + localhostip + ':3000/' + 'what_now',
        {
          method: 'GET',
          headers: {
            'device': deviceModel,
            'test_type': TEST_TYPE,
            'application_id': APP_ID,
            'framework': FRAMEWORK,
            'activity': ACTIVITY
          }
        }
      );
      benchmark = await response.text();
      hasCalled = true;
      console.log('executing: ' + benchmark);
      setLoading(false);


      //run program

      sendLogData();
    } catch (error) {
      console.error(error);
    }

  };

  whatNow();

  const sendLogData = async () => {
    try {
      console.log('http://' + localhostip + ':3000/' + 'logdata');
      let deviceModel = await DeviceInfo.getModel()
      const response = await fetch(
        'http://' + localhostip + ':3000/' + 'logdata',
        {
          method: 'GET',
          headers: {
            'device': deviceModel,
            'test_type': TEST_TYPE,
            'application_id': APP_ID,
            'framework': FRAMEWORK,
            'activity': ACTIVITY
          }
        }
      );
      const isDone = await response.text();
      done(isDone);
    } catch (error) {
      console.error(error);
    }
  };

  const done = async isDone => {
    if (isDone === 'done') {
      try {
        console.log('http://' + localhostip + ':3000/' + 'done');
        let deviceModel = await DeviceInfo.getModel()
        const response = await fetch(
          'http://' + localhostip + ':3000/' + 'done',
          {
            method: 'GET',
            headers: {
              'device': deviceModel,
              'test_type': TEST_TYPE,
              'application_id': APP_ID,
              'framework': FRAMEWORK,
              'activity': ACTIVITY
            }
          }
        );
      } catch (error) {
        console.error(error);
      }
    } else {
      console.log('not done?');
    }
  }
}